net/sched/sch_gkprio.c

```diff
@@ -0,0 +1,283 @@
+	int wlen, ret;
+	u16 prio, lp;
+	struct gkprio_sched_data *q;
+
+	wlen = skb_network_offset(skb);
+
+	switch (tc_skb_protocol(skb)) {
+	case htons(ETH_P_IP):
+		wlen += sizeof(struct iphdr);
+		if (!pskb_may_pull(skb, wlen))
+			goto drop;
+		prio = ipv4_get_dsfield(ip_hdr(skb)) >> 2;
+		break;
+
+	case htons(ETH_P_IPV6):
+		wlen += sizeof(struct ipv6hdr);
+		if (!pskb_may_pull(skb, wlen))
+			goto drop;
+		prio = ipv6_get_dsfield(ipv6_hdr(skb)) >> 2;
+		break;
+
+	default:
+		prio = 0;
+		break;
+	}
+
+	q = qdisc_priv(sch);
+	qdisc = &q->qdiscs[prio];
+
+	if (sch->q.qlen < q->max_limit) {
+		ret = __qdisc_enqueue_tail(skb, sch, qdisc);
+		if (ret < 0)
+			return ret;
+
+		/* Check to update highest and lowest priorities. */
+		if (prio > q->highest_prio)
+			q->highest_prio = prio;
+
+		if (prio < q->lowest_prio)
+			q->lowest_prio = prio;
+
+		sch->q.qlen++;
+		return ret;
+	}
+
+	/* If this packet is lowest priority, drop it. */
+	lp = q->lowest_prio;
+	if (prio <= lp)
+		return qdisc_drop(skb, sch, to_free);
+
+	/* Drop the packet at the tail of lowest priority qdisc. */
+	lp_qdisc = &q->qdiscs[lp];
+	if (!__qdisc_queue_drop_tail(sch, lp_qdisc, to_free))
+		goto drop;
+	qdisc_qstats_drop(sch);
+
+	ret = __qdisc_enqueue_tail(skb, sch, qdisc);
+	if (ret < 0)
+		return ret;
+
+	/* Check to update highest and lowest priorities. */
+	if (lp_qdisc->qlen == 0)
+		q->lowest_prio = calc_new_low_prio(q);
+
+	if (prio > q->highest_prio)
+		q->highest_prio = prio;
+
+	return ret;
+drop:
+	qdisc_drop(skb, sch, to_free);
+	return NET_XMIT_SUCCESS | __NET_XMIT_BYPASS;
+}
+
+static struct sk_buff *gkprio_dequeue(struct Qdisc *sch)
+{
+	struct gkprio_sched_data *q = qdisc_priv(sch);
+	struct sk_buff *skb;
+	struct qdisc_skb_head *hpq = &q->qdiscs[q->highest_prio];
+
+	skb = __qdisc_dequeue_head(hpq);
+	if (unlikely(!skb))
+		return NULL;
+
+	sch->q.qlen--;
+	qdisc_qstats_backlog_dec(sch, skb);
+	qdisc_bstats_update(sch, skb);
+
+	/* Update highest priority field. */
+	if (hpq->qlen == 0)
+		q->highest_prio = calc_new_high_prio(q);
+
+	return skb;
+}
+
+static int gkprio_init(struct Qdisc *sch, struct nlattr *opt)
+{
+	int prio;
+	struct gkprio_sched_data *q = qdisc_priv(sch);
+
+	if (!opt) {
+		q->max_limit = qdisc_dev(sch)->tx_queue_len;
+	} else {
+		struct tc_gkprio_qopt *ctl = nla_data(opt);
+
+		if (nla_len(opt) < sizeof(*ctl))
+			return -EINVAL;
+
+		q->max_limit = ctl->limit;
+	}
+
+	/* Initialise 64 queues, one for each possible priority. */
+	for (prio = 0; prio < GKPRIO_MAX_PRIORITY; prio++)
+		qdisc_skb_head_init(&q->qdiscs[prio]);
+
+	q->highest_prio = 0;
+	q->lowest_prio = GKPRIO_MAX_PRIORITY - 1;
+	return 0;
+}
+
+static int gkprio_dump(struct Qdisc *sch, struct sk_buff *skb)
+{
+ 	struct gkprio_sched_data *q = qdisc_priv(sch);
+	struct tc_gkprio_qopt opt;
+ 
+	opt.limit = q->max_limit;
+	opt.highest_prio = q->highest_prio;
+	opt.lowest_prio = q->lowest_prio;
+
+	if (nla_put(skb, TCA_OPTIONS, sizeof(opt), &opt))
+		goto nla_put_failure;
```

Comment:
Style: there is only one label (i.e. `nla_put_failure`) and it's only used here, so just `return -1;` here and drop the `goto` and label.